### PR TITLE
Adding Integration Tests for Dispute Edge Cases

### PR DESCRIPTION
**Pull request by jacktzhao**
`src/test/java/net/testudobank/tests/MvcControllerIntegTest.java
`
- Wrote an integration test to test the edge case where a reversing a deposit causes a customer to go above the overdraft limit of 1000$.
- Trying to reverse a deposit that causes the customer to go over the overdraft limit should result in the customer being directed to the welcome screen and not process the reversal.
- Wrote an integration test to test the edge case where reversing a deposit that causes a customer to fall into overdraft.
- Reversing a deposit that leads to an overdraft should apply the 2% interest rate to the overdraft balance.
- Wrote an integration test to test the edge case where reversing a deposit causes a customer to fall back into overdraft.
- In this case, reversing the deposit should leave the customer with the original overdraft balance and not apply the 2% interest fee.

`src/test/java/net/testudobank/tests/MvcControllerTest.java
`
- Removed mockito test for reversing a dispute that causes an overdraft 
- New integration tests cover the deleted cases

`src/main/java/net/testudobank/MvcController.java
`
- Added updated overdraft balance variable to display overdraft balance before the 2% interest fee is applied.
- Set overdraft balance to updated overdraft balance in the case where reversing a dispute causes a customer to fall back into overdraft balance

### CODE REVIEW

File: src/main/java/net/testudobank/MvcController.java [hunk 1]
```diff
@@ -149,7 +152,13 @@ private static int convertDollarsToPennies(double dollarAmount) {
     return (int) (dollarAmount * 100);
   }
 
-  //// HTML POST HANDLERS ////
+  // Converts LocalDateTime to Date variable
+  private static Date convertLocalDateTime(LocalDateTime ldt){
```
Comment (adithyasolai) [inline src/main/java/net/testudobank/MvcController.java:156]: change method name to `convertLocalDateTimeToDate`.

File: src/main/java/net/testudobank/MvcController.java [hunk 2]
```diff
@@ -395,10 +404,18 @@ public String submitDispute(@ModelAttribute("user") User user) {
       // If reversing a deposit puts customer back in overdraft
       if (reversalAmountInPennies > userBalanceInPennies){
         // check if the reversed deposit helped pay off overdraft balance
-        List<Map<String,Object>> overdraftLogs = TestudoBankRepository.getOverdraftLogs(jdbcTemplate, userID, (String)logToReverse.get("Timestamp"));
+        String dateTime = SQL_DATETIME_FORMATTER.format(convertLocalDateTime((LocalDateTime)logToReverse.get("Timestamp")));
```
Comment (adithyasolai) [inline src/main/java/net/testudobank/MvcController.java:407]: Maybe rename this variable to something like `datetimeOfReversedDeposit` for more clarity.

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java [hunk 3]
```diff
@@ -704,4 +704,224 @@ public void testFrozenAccount() throws SQLException, ScriptException, Interrupte
...
+   * Trying to reverse a deposit that causes the customer to go over the overdraft limit
+   * should result in the customer being directed to the welcome screen and not process 
+   * the reversal.
+   * 
+   * @throws SQLException
+   * @throws ScriptException
+   * @throws InterruptedException
+   */
+  @Test
+  public void testReverseDepositExceed() throws SQLException, ScriptException, InterruptedException {
```
Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:722]: rename method name to `testReverseDepositExceedsOverdraftLimit`

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java [hunk 4]
```diff
@@ -704,4 +704,224 @@ public void testFrozenAccount() throws SQLException, ScriptException, Interrupte
     assertEquals(2, transactionHistoryTableData.size());
   }
 
+  /**
+   * Verifies the transaction dispute feature on a reversal of a deposit that 
+   * causes a customer to above the overdraft limit.
```
Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:709]: Just a grammar nitpick: "causes a customer to exceed the overdraft limit"

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java [hunk 5]
```diff
@@ -704,4 +704,224 @@ public void testFrozenAccount() throws SQLException, ScriptException, Interrupte
...
+    // send Dispute request
+    String responsePage = controller.submitDispute(customer1ReversalFormInputs);
+    assertEquals("welcome", responsePage);
+
+
+    // fetch transaction data from the DB in chronological order
+    // the more recent transaction should be the Withdraw, and the older transaction should be the Deposit
+    List<Map<String,Object>> transactionHistoryTableData = jdbcTemplate.queryForList("SELECT * FROM TransactionHistory ORDER BY Timestamp ASC;");
+    Map<String,Object> customer1DepositTransactionLog = transactionHistoryTableData.get(0);
+    Map<String,Object> customer1WithdrawTransactionLog = transactionHistoryTableData.get(1);
```
Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:776]: Before fetching the transaction logs, I would do a quick check to ensure that the TransactionHistory result only has 2 entries. If the reversed Deposit did actually go through, then it would be logged as a Withdraw in the TransactionHistory, which would make the query result here give 3 entries back. I know that you already checked that the responsePage is `"welcome"` above, but there are many invalid input/state cases that can cause a `"welcome"` response. This extra check could help us hone in on the specific invalid input/state case we are testing for.

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java [hunk 6]
```diff
@@ -704,4 +704,224 @@ public void testFrozenAccount() throws SQLException, ScriptException, Interrupte
...
+    Map<String,Object> customer1Data = customersTableData.get(0);
+    assertEquals(0, (int)customer1Data.get("Balance"));
+    
+    // verify that customer1's Overdraft balance is equal to the remaining withdraw amount with interest applied
+    // (convert to pennies before applying interest rate to avoid floating point roundoff errors when applying the interest rate)
+    int CUSTOMER1_ORIGINAL_BALANCE_IN_PENNIES = MvcControllerIntegTestHelpers.convertDollarsToPennies(CUSTOMER1_BALANCE);
+    int CUSTOMER1_AMOUNT_TO_WITHDRAW_IN_PENNIES = MvcControllerIntegTestHelpers.convertDollarsToPennies(CUSTOMER1_AMOUNT_TO_WITHDRAW);
+    int CUSTOMER1_AMOUNT_TO_DEPOSIT_IN_PENNIES = MvcControllerIntegTestHelpers.convertDollarsToPennies(CUSTOMER1_AMOUNT_TO_DEPOSIT);
+    int CUSTOMER1_AMOUNT_TO_REVERSE_IN_PENNIES = MvcControllerIntegTestHelpers.convertDollarsToPennies(CUSTOMER1_AMOUNT_TO_DEPOSIT);
+    int CUSTOMER1_EXPECTED_OVERDRAFT_BALANCE_BEFORE_INTEREST_IN_PENNIES = CUSTOMER1_ORIGINAL_BALANCE_IN_PENNIES + CUSTOMER1_AMOUNT_TO_DEPOSIT_IN_PENNIES + CUSTOMER1_AMOUNT_TO_WITHDRAW_IN_PENNIES - CUSTOMER1_AMOUNT_TO_REVERSE_IN_PENNIES;
```
Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:857]: I really like how clear this code is to read. It's OK to be excessive in integ tests because people only read the code when tests fail. In the Controller logic code, we would prefer to be more elegant and minimize repetition since that code is read much more frequently.

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java [hunk 7]
```diff
@@ -704,4 +704,224 @@ public void testFrozenAccount() throws SQLException, ScriptException, Interrupte
...
+    // Prepare Deposit Form to Deposit $100 to customer 1's account.
+    double CUSTOMER1_AMOUNT_TO_DEPOSIT = 100; // user input is in dollar amount, not pennies.
+    User customer1DepositFormInputs = new User();
+    customer1DepositFormInputs.setUsername(CUSTOMER1_ID);
+    customer1DepositFormInputs.setPassword(CUSTOMER1_PASSWORD);
+    customer1DepositFormInputs.setAmountToDeposit(CUSTOMER1_AMOUNT_TO_DEPOSIT);
+
+    // store timestamp of when Deposit request is sent to verify timestamps in the TransactionHistory table later
+    LocalDateTime timeWhenDepositRequestSent = MvcControllerIntegTestHelpers.fetchCurrentTimeAsLocalDateTimeNoMilliseconds();
+    System.out.println("Timestamp when Deposit Request is sent: " + timeWhenDepositRequestSent);
```
Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:903]: Can remove this because we don't do any timestamp checks in this integ test.

File: src/main/java/net/testudobank/MvcController.java [hunk 8]
```diff
@@ -395,10 +404,18 @@ public String submitDispute(@ModelAttribute("user") User user) {
       // If reversing a deposit puts customer back in overdraft
       if (reversalAmountInPennies > userBalanceInPennies){
         // check if the reversed deposit helped pay off overdraft balance
```
Comment (adithyasolai) [inline src/main/java/net/testudobank/MvcController.java:406]: Add to this comment saying something like "if it did, do not re-apply the interest rate after the reversal of the deposit since the customer was already in overdraft."

File: src/main/java/net/testudobank/MvcController.java [hunk 9]
```diff
@@ -395,10 +404,18 @@ public String submitDispute(@ModelAttribute("user") User user) {
...
-        List<Map<String,Object>> overdraftLogs = TestudoBankRepository.getOverdraftLogs(jdbcTemplate, userID, (String)logToReverse.get("Timestamp"));
+        String dateTime = SQL_DATETIME_FORMATTER.format(convertLocalDateTime((LocalDateTime)logToReverse.get("Timestamp")));
+        List<Map<String,Object>> overdraftLogs = TestudoBankRepository.getOverdraftLogs(jdbcTemplate, userID, dateTime);
+
+        // fetch updated overdraft balance and correct interest rate
```
Comment (adithyasolai) [inline src/main/java/net/testudobank/MvcController.java:410]: change this comment to `// fetch updated overdraft balance with extra interest rate applied`

File: src/main/java/net/testudobank/MvcController.java [hunk 10]
```diff
@@ -395,10 +404,18 @@ public String submitDispute(@ModelAttribute("user") User user) {
...
+        double updatedOverdraftBalanceInPennies = TestudoBankRepository.getCustomerOverdraftBalanceInPennies(jdbcTemplate, userID);
+        double newOverdraftBalanceInPennies = updatedOverdraftBalanceInPennies / 1.02;
```
Comment (adithyasolai) [inline src/main/java/net/testudobank/MvcController.java:412]: Add a comment above this line saying `// reverse extra application of interest rate since customer was already in overdraft`.

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java [hunk 11]
```diff
@@ -704,4 +704,224 @@ public void testFrozenAccount() throws SQLException, ScriptException, Interrupte
...
+   * Reversing a deposit that causes a customer to fall into overdraft should 
+   * make the customer go into overdraft and apply the 2% interest fee on 
+   * the overdraft balance.
+   * 
+   * @throws SQLException
+   * @throws ScriptException
+   * @throws InterruptedException
+   */
+  @Test
+  public void testReverseDepositOverdraft() throws SQLException, ScriptException, InterruptedException {
```
Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:801]: Change method name to `testReverseDepositCausesOverdraft`

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java [hunk 12]
```diff
@@ -704,4 +704,224 @@ public void testFrozenAccount() throws SQLException, ScriptException, Interrupte
...
+   * Reversing a deposit that causes a customer to fall back into an overdraft balance 
+   * greater than 0 should put the customer back into the original overdraft balance. 
+   * The 2% interest rate should not be re-applied.
+   * 
+   * @throws SQLException
+   * @throws ScriptException
+   * @throws InterruptedException
+   */
+  @Test
+  public void testReverseDepositRePayOverdraft() throws SQLException, ScriptException, InterruptedException {
```
Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:877]: Change method name to `testReverseDepositThatRepaysOverdraft`

**Review by adithyasolai — CHANGES_REQUESTED** [review]
Good job!

Most of my comments are related to changing variable names + comments and removing unneeded code to improve readability.

**Review by adithyasolai — APPROVED** [review]
Pushed my own small changes for nitpicky things. Merging now.
